clay/random.clj

```diff
@@ -1,11 +1,2259 @@
+                           [(kind/code (str k)) v]))
+                    (into []))})
+
+;; KDE kernels
+
+^:kindly/hide-code
+(def kde-kernels (:kde k/kernel-list))
+
+(kind/table
+ (->> kde-kernels
+      (mapv (fn [k]
+              [(kind/hiccup
+                [:dl
+                 [:dt (kind/code (str k))]
+                 [:dd (kind/table [(mapv gg/->image (gg/dgraph (r/distribution :continuous-distribution {:data [-2 -2 -2 -1 0 1 2 -1 0 1 2 0 1 2 1 2 2]
+                                                                                                         :kde k :steps 100}) {:pdf [-3 3]}))])]])]))))
+
+
+;; #### Empirical
+
+;; Empirical distribution calculates PDF, CDF and iCDF from a histogram.
+
+;; * Name: `:empirical`
+;; * Default parameters
+;;    * `:data`, samples, sequence of numbers
+;;    * `:bin-count`, number of bins for histogram: 10% of the size of the data
+;; * [wiki](https://en.wikipedia.org/wiki/Empirical_distribution_function), [source](https://commons.apache.org/proper/commons-math/javadocs/api-3.6.1/org/apache/commons/math3/random/EmpiricalDistribution.html)
+
+(def empirical-distr (r/distribution :empirical {:data random-data}))
+
+(utls/dgraph-table
+ [["default", "bin-count=10"]
+  [(mapv gg/->image (gg/dgraph empirical-distr {:pdf [-5 5]}))
+   (mapv gg/->image (gg/dgraph (r/distribution :empirical {:data random-data :bin-count 10}) {:pdf [-5 5]}))]])
+
+(utls/examples-note
+ (r/sample empirical-distr)
+ (r/sample empirical-distr)
+ (r/pdf empirical-distr 0.0)
+ (r/cdf empirical-distr 0.0)
+ (map (partial r/icdf empirical-distr) [0.0001 0.5 0.9999])
+ (r/lower-bound empirical-distr)
+ (r/upper-bound empirical-distr)
+ (r/mean empirical-distr)
+ (r/variance empirical-distr))
+
+;; #### Discrete
+
+;; * Default parameters
+;;    * `:data`, sequence of numbers (integers/longs or doubles)
+;;    * `:probabilities`, optional, probabilities or weights
+;; * [wiki](https://en.wikipedia.org/wiki/Probability_distribution#Discrete_probability_distribution), [source1](https://commons.apache.org/proper/commons-math/javadocs/api-3.6.1/org/apache/commons/math3/distribution/EnumeratedRealDistribution.html), [source2](https://commons.apache.org/proper/commons-math/javadocs/api-3.6.1/org/apache/commons/math3/distribution/EnumeratedIntegerDistribution.html)
+
+;; Please note: data can contain duplicates.
+
+;; There are four discrete distributions:
+;; * `:enumerated-int` for integers, backed by Apache Commons Math
+;; * `:enumerated-real` for doubles, backed by Apache Commons Math
+;; * `:integer-discrete-distribution` - for longs, custom implementation
+;; * `:real-discrete-distribution` - for doubles, custom implementation
+
+;; Please note:
+
+;; * Apache Commons Math implementation have some issues with iCDF.
+;; * `:integer-discrete-distribution` is backed by `clojure.data.int-map`
+
+;; ##### Doubles
+
+(def data-doubles (repeatedly 100 #(m/sq (m/approx (r/drand 2.0) 1))))
+
+(kind/table
```

Comment:
Let's make it in two rows for clarity.

Reply:
Yes, both sets of charts are presented vertically.